File: 卡码/代码随想录/1/06-二叉树/Js/04-二叉搜索树属性/LeetCode_0501.js
//501. 二叉搜索树中的众数
import TreeNode from "../TreeNode.js";

const findMode = function(root) {
  let pre=null;
  let res=new Array();
  let cnt=0,maxCnt=0;
  const traversal=root=>{
    if(!root) return;
    traversal(root.left);
    if(!pre||pre.val!=root.val) cnt=1;
    else cnt++;
    pre=root;
    if(cnt==maxCnt) res.push(root.val);
    else if(cnt>maxCnt){
      maxCnt=cnt;
      res=[root.val];
    }
    traversal(root.right);
    return;
  };
  traversal(root);
  return res;
};

File: 卡码/代码随想录/1/06-二叉树/Python/04-二叉搜索树属性/LeetCode_0501.py
#501. 二叉搜索树中的众数
from typing import *
from ..TreeNode import *

class Solution:
  def __init__(self):
    self.pre:TreeNode=None
    self.res:List[int]=[]
    self.count:int=0
    self.maxCount:int=0
  def traversal(self,root:Optional[TreeNode]):
    if not root:
      return
    self.traversal(root.left)
    if not self.pre or self.pre.val!=root.val:
      self.count=1
    else:
      self.count+=1
    self.pre=root
    if self.count==self.maxCount:
      self.res.append(root.val)
    elif self.count>self.maxCount:
      self.maxCount=self.count
      self.res=[root.val]
    self.traversal(root.right)
    return
  def findMode(self, root: Optional[TreeNode]) -> List[int]:
    self.traversal(root)
    return self.res
